# Step 2: Loading the Case Study Data in a Jupyter Notebook

# Starting with Jupyter notebooks

There are two kinds of cells: markdown and 'code' cells. This one is marked as markdown. The next one is a `markdown` cell, followed by a 'code' cell.

# The Business context of the data

Our consultancy firm is contracted by a credit card client. The client have offered us a data set with data from account holders. Each row corresponds to one customer's account data. The data contains monthly records, over a period of 6 months. Data from 30,000 acount holders are provided. The data are classified according to whether an account owner has defaulted, after a six month period. In practice this means that the account holder did not make the minimum required payment. 

# The goal for data analytics

The client wants us to develop a model to predic whether in the month after the six-month period of the historical data, e account holder will default or not. 


# The data set

The data set is a modified version of the following: 
https://archive.ics.uci.edu/ml/datasets/default+of+credit+card+clients. 

This is taken from the UCI Machine Learning Repository, a public repository of bencmarking datasets. .
Have a look at that web page to find out more abot the data characteristics. Can you summarise the key characteristics?

Next, let's turn our attention to the code needed to load the data. 

In [1]:
import pandas as pd

With the first command we imported the pandas data management library. 

Next we will use it to load our first data set. 

Data loading is done in the next code cell. 

In [2]:
import os 

In [3]:
#path = 'C:\Python\Code\SIO\Practicals\Practicals 1\Data'
#you may set your own path here

In [4]:
#df = pd.read_excel("default_of_credit_card_clients.xls") 
#df = pd.read_csv('default_of_credit_card_clients.csv',index_col='ID')
df = pd.read_csv('default_of_credit_card_clients.csv')

If we have a separate directory Data, then we need to set the path. Otherwise we just read the file from the current directory. The file is available on Brightspace. Here we read the 'ID' as attribute, so we can manipulate it. If we wanted to register the 'ID' values as the actual ID, we would have useed the index_col='ID' option above. 
We start our exploratory analysis by inspecting the size of the data file

In [6]:
df.shape

(30000, 26)

So this file has 30000 records. We have 26 attributes but we have loaded also the ID as attribute so as to be able to include the ID in the data manipulates. 

We can also execute direct value assignment commands and view their outcomes, as below

# Step 3: Check Data Integrity

The data set contains monthly credit card account data, for a period of six months. Let's perform a quality check to ensure we have the data for the accounts as expected. The account ID distinguishes one account from the other. We can check unique IDs with Pandas with the function `.nunique()`. But first let's check our data structure. To do so we first build an Index of he different columns in the table. We will use the .columns method of the pandas DataFrame to see the column names. 

In [7]:
df.columns

Index(['Unnamed: 0', 'ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE',
       'PAY_1', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1',
       'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6',
       'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

We then check the column headings

In [8]:
df.head()

Unnamed: 0             ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  \
0           0  798fc410-45c1      20000    2          2         1   24     2   
1           1  8a8c8f3b-8eb4     120000    2          2         2   26    -1   
2           2  85698822-43f5      90000    2          2         2   34     0   
3           3  0737c11b-be42      50000    2          2         1   37     0   
4           4  3b7f77cc-dbc0      50000    1          2         1   57    -1   

   PAY_2  PAY_3  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  PAY_AMT2  \
0      2     -1  ...          0          0          0         0       689   
1      2      0  ...       3272       3455       3261         0      1000   
2      0      0  ...      14331      14948      15549      1518      1500   
3      0      0  ...      28314      28959      29547      2000      2019   
4      0     -1  ...      20940      19146      19131      2000     36681   

   PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  default payment next month  
0         0         0         0         0                           1  
1      1000      1000         0      2000                           1  
2      1000      1000      1000      5000                           0  
3      1200      1100      1069      1000                           0  
4     10000      9000       689       679                           0  

[5 rows x 26 columns]

And now we check the unique IDs

In [9]:
df['ID'].nunique()

29687

That's interesting! We have fewer unique IDs than rows, so we clearly have duplicates! We can check the number of occurences of each ID by counting them, as follows: 

In [10]:
id_counts = df['ID'].value_counts()
id_counts.head()

ad23fe5c-7b09    2
1fb3e3e6-a68d    2
89f8f447-fca8    2
7c9b7473-cc2f    2
90330d02-82d9    2
Name: ID, dtype: int64

In [11]:
id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

So now we know that we have 29374 IDs which appear only once but we also hae 313 which appear twice!

In the previous, most comments have now been taken out for compactness. 

# Filtering the data

Before we progress with data manipulation, let's focus on how we can filter the data using a range of 'masks' - these are Boolean operations on the data. we will first use synthetic data to illustrate this. We will import the main numerical processing packge, namely NumPy. This has among other capabilities for generating random numbers. 


In [12]:
import numpy as np
from numpy.random import default_rng

We will define 'seed' numbers for the random data generation. 

In [13]:
rg = default_rng(12345)

And now we will define the range values of the data. Note that the upper value is never reached. 

In [14]:
random_integers = rg.integers(low=1,high=5,size=100)

We now look at the first 5 rows

In [15]:
random_integers[:5]

array([3, 1, 4, 2, 1], dtype=int64)

Next we are interested in identifing the locations of all random integers equal to 3

In [16]:
is_equal_to_3 = random_integers == 3

Let's just focus on the first rows again

In [17]:
is_equal_to_3[:5]

array([ True, False, False, False, False])

But how many rows have a random number with value that equals to 3?

In [19]:
sum(is_equal_to_3)

31

And we can verify this below

In [20]:
random_integers[is_equal_to_3]

array([3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3], dtype=int64)

Now that we have familiarised with using bolean masks for filtering data, let's continue with our data integrity check. 

# Step 4: Continue with Data Integrity

We will now examine in more detail the duplicate IDs. To do so, let's create a Boolean mask to locate the duplicate IDs. We name this 'mask dupe_mask'. Let's display the first five elements.

In [21]:
dupe_mask = id_counts == 2

In [22]:
dupe_mask[0:5]

ad23fe5c-7b09    True
1fb3e3e6-a68d    True
89f8f447-fca8    True
7c9b7473-cc2f    True
90330d02-82d9    True
Name: ID, dtype: bool

You can change the '5' in the command above with another number and observe a different number of records. Let's keep it '5'now and inspect the duplicate ID records. 

In [23]:
id_counts.index[0:5]

Index(['ad23fe5c-7b09', '1fb3e3e6-a68d', '89f8f447-fca8', '7c9b7473-cc2f',
       '90330d02-82d9'],
      dtype='object')

To be able to access and inspect these records, let's store their ID numbers in a variable named 'dupe_ids' and inspect the first 10 such records:  

In [24]:
dupe_ids = id_counts.index[dupe_mask]

In [25]:
dupe_ids[:10]

Index(['ad23fe5c-7b09', '1fb3e3e6-a68d', '89f8f447-fca8', '7c9b7473-cc2f',
       '90330d02-82d9', '2a793ecf-05c6', '75938fec-e5ec', '7be61027-a493',
       'a3a5c0fc-fdd6', 'b44b81b2-7789'],
      dtype='object')

Let's double check the cardinality of the dupe_ids variable (i.e. how many records it contains). We use 'len' for this: 

In [26]:
dupe_ids = list(dupe_ids)
len(dupe_ids)

313

That's right - as expected. So we focus again on the first 5. 

In [27]:
dupe_ids[0:5]

['ad23fe5c-7b09',
 '1fb3e3e6-a68d',
 '89f8f447-fca8',
 '7c9b7473-cc2f',
 '90330d02-82d9']

Now that we have selected the records with the duplicate IDs, we wish to inspect their data in more detail. Is there anything different in the values of these records which have identical IDs? Our approach to do so is this: we first focus on the data records with the first 3 duplicate IDs, as a sample. First we focus on finding the rows which contain these 3 IDs. The first three IDs are dupe_ids[0:3]. 

We will use the '.isin' method to create another logical mask. We will pass the earlier selected list of IDs through this mask and apply this on the larger dataframe to display the rows that have this ID. We use the .loc method for locating these IDs. The '.isin' method is NESTED within the .loc statement: this indexes the datafame to select the location of all rows containing a True value in the logical mask. The : in the second argument of the loc implies that all columns will be seleted. So in this way we filter the Dataframe so as to view all the columns for the first three duplicate IDs. This is not very comlex filtering but still it is a sign of the compactness of the Python language that a single command with this nested structure produces this outcome. 

In [28]:
df.loc[df['ID'].isin(dupe_ids[0:3]),:]

Unnamed: 0             ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE  \
5033         5033  89f8f447-fca8     320000    2          2         1   32   
5133         5133  89f8f447-fca8          0    0          0         0    0   
16727       16727  1fb3e3e6-a68d      80000    1          2         2   33   
16827       16827  1fb3e3e6-a68d          0    0          0         0    0   
29685       29685  ad23fe5c-7b09      50000    1          3         1   32   
29785       29785  ad23fe5c-7b09          0    0          0         0    0   

      PAY_1  PAY_2  PAY_3  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  PAY_AMT1  \
5033      0      0      0  ...     169371     172868     150827      8000   
5133      0      0      0  ...          0          0          0         0   
16727     2      2      0  ...      27394      29922      31879         0   
16827     0      0      0  ...          0          0          0         0   
29685     0      0      0  ...      12882       8131       3983      3000   
29785     0      0      0  ...          0          0          0         0   

       PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
5033       8000      5500      6100      6000      5000   
5133          0         0         0         0         0   
16727      2000      2000      3000      2600         0   
16827         0         0         0         0         0   
29685      2871      1000       163      3983      3771   
29785         0         0         0         0         0   

       default payment next month  
5033                            0  
5133                            0  
16727                           1  
16827                           0  
29685                           1  
29785                           0  

[6 rows x 26 columns]

This gets interesting. For each pair of rows with duplicate IDs, one row contains potentially valid data, but the other one only contains 0s! It is most likley that a data clearning process would be expected to delete all such eroneous data. You can't for example have an age of 0 or a credit limit of 0 and stil be part of the data processing! 

We can try to delete all rows with data values of zero (excluding the ID attribute). First we inspect the unique values in the data to see if there are any other problems.

In [29]:
column_names = df.columns.values.tolist()

First we get the attribute names

In [30]:
print(column_names)

['Unnamed: 0', 'ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_1', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6', 'default payment next month']


Next we count the unique values for each attribute

In [31]:
count_of_unique_values = df.nunique()
print(count_of_unique_values)

Unnamed: 0                    30000
ID                            29687
LIMIT_BAL                        82
SEX                               3
EDUCATION                         7
MARRIAGE                          4
AGE                              57
PAY_1                            12
PAY_2                            11
PAY_3                            11
PAY_4                            11
PAY_5                            10
PAY_6                            10
BILL_AMT1                     22510
BILL_AMT2                     22146
BILL_AMT3                     21822
BILL_AMT4                     21350
BILL_AMT5                     20831
BILL_AMT6                     20417
PAY_AMT1                       7890
PAY_AMT2                       7847
PAY_AMT3                       7463
PAY_AMT4                       6901
PAY_AMT5                       6857
PAY_AMT6                       6895
default payment next month        2
dtype: int64


Next we inspect the unique values for each attribute

In [32]:
for index in column_names :
    uniquenamesincolumn = df[index].unique() 
    print(uniquenamesincolumn)

[    0     1     2 ... 29997 29998 29999]
['798fc410-45c1' '8a8c8f3b-8eb4' '85698822-43f5' ... '95cdd3e7-4f24'
 '00d03f02-04cd' '15d69f9f-5ad3']
[  20000  120000   90000   50000  500000  100000  140000  200000  260000
  630000   70000  250000  320000  360000  180000  130000  450000   60000
  230000  160000  280000   10000   40000  210000  150000  380000  310000
  400000   80000  290000  340000  300000   30000  240000  470000  480000
  350000  330000  110000  420000  170000  370000  270000  220000  190000
       0  510000  460000  440000  410000  490000  390000  580000  600000
  620000  610000  700000  670000  680000  430000  550000  540000 1000000
  530000  710000  560000  520000  750000  640000   16000  570000  590000
  660000  720000  327680  740000  800000  760000  690000  650000  780000
  730000]
[2 1 0]
[2 1 3 5 0 4 6]
[1 2 3 0]
[24 26 34 37 57 29 23 28 35 51 41 30 49 39 40 27 47 33 32 54 58 22 25 31
 46 42 43 45 56 44 53 38 63 36 52 48 55 60 50  0 75 61 73 59 21 67 66 62
 70 72 6

The Attribute 'PAY_1' appears to have a 'Not available' value in at least one of the records. We will delete these records next. 

In [36]:
df = df[df.PAY_1 != 'Not available']

Now we will examine our data records again

In [37]:
count_of_unique_values = df.nunique()
print(count_of_unique_values)

Unnamed: 0                    26979
ID                            26704
LIMIT_BAL                        81
SEX                               3
EDUCATION                         7
MARRIAGE                          4
AGE                              57
PAY_1                            11
PAY_2                            11
PAY_3                            11
PAY_4                            11
PAY_5                            10
PAY_6                            10
BILL_AMT1                     20519
BILL_AMT2                     20165
BILL_AMT3                     19852
BILL_AMT4                     19472
BILL_AMT5                     19002
BILL_AMT6                     18642
PAY_AMT1                       7425
PAY_AMT2                       7361
PAY_AMT3                       7025
PAY_AMT4                       6463
PAY_AMT5                       6447
PAY_AMT6                       6471
default payment next month        2
dtype: int64


We have confirmed that there is one unique value less in PAY_1

In [38]:
for index in column_names :
    uniquenamesincolumn = df[index].unique() 
    print(uniquenamesincolumn)

[    0     1     2 ... 29997 29998 29999]
['798fc410-45c1' '8a8c8f3b-8eb4' '85698822-43f5' ... '95cdd3e7-4f24'
 '00d03f02-04cd' '15d69f9f-5ad3']
[ 20000 120000  90000  50000 500000 100000 140000 200000 260000 630000
  70000 250000 360000 180000 130000 450000  60000 230000 160000 280000
  10000  40000 210000 150000 380000 310000 400000  80000 290000 340000
 300000  30000 240000 470000 480000 350000 330000 110000 420000 170000
 370000 270000 220000 320000 190000      0 510000 460000 440000 410000
 490000 390000 580000 600000 620000 610000 700000 670000 680000 430000
 550000 540000 530000 710000 520000 750000 640000 570000 590000 660000
 560000 720000 327680 740000 800000 760000 690000 650000 780000 730000
  16000]
[2 1 0]
[2 1 3 5 0 4 6]
[1 2 3 0]
[24 26 34 37 57 29 23 28 35 51 41 30 49 39 40 27 33 32 54 58 22 25 31 46
 42 43 45 56 44 53 38 63 36 52 47 48 55 60 50  0 75 61 73 59 21 67 66 62
 70 72 64 65 71 68 69 79 74]
['2' '-1' '0' '-2' '1' '3' '4' '8' '7' '5' '6']
[ 2  0 -1 -2  3  5  7

Indeed, we don't see the 'Not available' value any more. 

However, the fact that we had this issue, impacted on the type of variable that PAY_1 was. We inspect the data types. 

In [40]:
df.dtypes

Unnamed: 0                     int64
ID                            object
LIMIT_BAL                      int64
SEX                            int64
EDUCATION                      int64
MARRIAGE                       int64
AGE                            int64
PAY_1                         object
PAY_2                          int64
PAY_3                          int64
PAY_4                          int64
PAY_5                          int64
PAY_6                          int64
BILL_AMT1                      int64
BILL_AMT2                      int64
BILL_AMT3                      int64
BILL_AMT4                      int64
BILL_AMT5                      int64
BILL_AMT6                      int64
PAY_AMT1                       int64
PAY_AMT2                       int64
PAY_AMT3                       int64
PAY_AMT4                       int64
PAY_AMT5                       int64
PAY_AMT6                       int64
default payment next month     int64
dtype: object

PAY_1 is the only attribute which does not appear to be of Integer type. We will convert it first to Integer type. 

In [44]:
df = df.astype({'PAY_1':int})

In [45]:
df.dtypes

Unnamed: 0                     int64
ID                            object
LIMIT_BAL                      int64
SEX                            int64
EDUCATION                      int64
MARRIAGE                       int64
AGE                            int64
PAY_1                          int32
PAY_2                          int64
PAY_3                          int64
PAY_4                          int64
PAY_5                          int64
PAY_6                          int64
BILL_AMT1                      int64
BILL_AMT2                      int64
BILL_AMT3                      int64
BILL_AMT4                      int64
BILL_AMT5                      int64
BILL_AMT6                      int64
PAY_AMT1                       int64
PAY_AMT2                       int64
PAY_AMT3                       int64
PAY_AMT4                       int64
PAY_AMT5                       int64
PAY_AMT6                       int64
default payment next month     int64
dtype: object

Not all attributed apart from ID are of Integer type. However, they are of unecessary high presicion (int64). It does not really matter for the present example, but it is possible to convert them all to the same type of integer. 

In [46]:
df = df.astype({col: 'int32' for col in df.select_dtypes('int64').columns})

In [47]:
df.dtypes

Unnamed: 0                     int32
ID                            object
LIMIT_BAL                      int32
SEX                            int32
EDUCATION                      int32
MARRIAGE                       int32
AGE                            int32
PAY_1                          int32
PAY_2                          int32
PAY_3                          int32
PAY_4                          int32
PAY_5                          int32
PAY_6                          int32
BILL_AMT1                      int32
BILL_AMT2                      int32
BILL_AMT3                      int32
BILL_AMT4                      int32
BILL_AMT5                      int32
BILL_AMT6                      int32
PAY_AMT1                       int32
PAY_AMT2                       int32
PAY_AMT3                       int32
PAY_AMT4                       int32
PAY_AMT5                       int32
PAY_AMT6                       int32
default payment next month     int32
dtype: object

Now we would like to identify the data records which have zero values for all attributes (apart from ID)

In [73]:
df_zero_mask = df == 0

In [79]:
feature_zero_mask = df_zero_mask.iloc[:,2:].all(axis=1)

In [80]:
sum(feature_zero_mask)

315

That's good. This is just 2 higher than the 313 duplicate IDs we have identified earlier. But even with non-duplicate IDs, having a data record with all 0s is not useful. So if we delete all 315 records with zero values, there is a chance that we may get rid of the records with duplicate IDs in a meaningful way. Let's do just that. First we create a new dataframe, called df_clean_1. This creates a copy of the original dataframe, but masked to NOT include (note the ~ symbol, which stands for negation) the records which contain the zero values) .

Now we need to select the record IDs of the above records. 

In [81]:
df_clean_1 = df.loc[~feature_zero_mask,:].copy()

In [82]:
df_clean_1.shape

(26664, 26)

In [83]:
df_clean_1['ID'].nunique()

26664

So this appears to have worked. You can save the "cleaner" data set for further processing next. 

In [84]:
df_clean_1.to_csv('df_clean_1.csv', index=False)